File: transform_queue.py
from stack import Stack

class StackedQueue(object):
	def __init__(self):
		self.stack1 = Stack()
		self.stack2 = Stack()
		self.is_pushed = False
	
	def _move_to_stack1(self):
		if not self.is_pushed:
			while not self.stack2.is_empty():
				self.stack1.push(self.stack2.pop())
		self.is_pushed = True

	def _move_to_stack2(self):
		if self.is_pushed:
			while not self.stack1.is_empty():
				self.stack2.push(self.stack1.pop())
		self.is_pushed = False

	def push(self, data):
		self._move_to_stack1()
		self.stack1.push(data)

	def pop(self):
		self._move_to_stack2()
		return self.stack2.pop()

	def is_empty(self):
		return self.stack1.is_empty() and self.stack2.is_empty()

	def peek(self):
		self._move_to_stack2()
		return self.stack2.peek()

def main():
	queue = StackedQueue()
	print("is empty?")
	print(queue.is_empty())
	print(queue.pop())
	print(queue.pop())
	queue.push(3)
	queue.push(6)
	queue.push(8)
	queue.push(0)
	queue.push(4)
	queue.push(5)
	print(queue.is_empty())
	print(queue.pop())
	print(queue.pop())
	print(queue.pop())
	print(queue.pop())
	# print(queue.pop())
	# print(queue.pop())
	queue.push(7)
	queue.push(9)
	print(queue.pop())
	print(queue.pop())
	print(queue.pop())
	print(queue.pop())
	queue.push(8)
	queue.push(0)
	queue.push(4)
	queue.push(5)
	print(queue.pop())
	print(queue.pop())
	print(queue.pop())
	print(queue.pop())
	queue.push(101)
	queue.push(102)
	print(queue.pop())
	queue.push(103)
	print(queue.pop())
	queue.push(104)
	print(queue.pop())
	print(queue.peek())
	print(queue.peek())
	print(queue.peek())
	queue.push(101)
	print(queue.peek())
	queue.push(101)
	print(queue.pop())
	print(queue.pop())
	print(queue.pop())
	print(queue.pop())
	# data = queue.peek()
	# print(data)
	# print("is empty?")

	# print(queue.is_empty())

if __name__ == "__main__":
	main()

File: transform_queue.cpp
#include <iostream>

using namespace std;

#define NULL_VALUE -1

class Node{
public:
	Node(){
		next = NULL;
	}
	int data;
	Node* next;
};

class Stack{
public:
	Stack(){
		head = NULL;
	}
	int push(int);
	int pop();
	bool isEmpty();
	int peek();
private:
	Node* head;
};

int Stack::push(int data){
	Node* node = new Node;
	node->data = data;
	node->next = head;
	head = node;
}
int Stack::pop(){				
	if(isEmpty()){
		return NULL_VALUE;
	}
	int data;
	data = head->data;
	head = head->next;
	return data;
}
bool Stack::isEmpty(){
	return head == NULL;
}
int Stack::peek(){
	if(head != NULL){
		return head->data;
	}
	return NULL_VALUE; 
}

class StackedQueue{
public:
	StackedQueue(){
		stack1 = new Stack();
		stack2 = new Stack();
		isPushed = false;
	}
	int push(int);
	int pop();
	bool isEmpty();
	int peek();
private:
	int moveToStack1();
	int moveToStack2();
	Stack* stack1;
	Stack* stack2;
	bool isPushed;
};

int StackedQueue::moveToStack1(){
	if(!isPushed){
		while(!stack2->isEmpty()){
			stack1->push(stack2->pop());
		}
	isPushed = true;
	}
}
int StackedQueue::moveToStack2(){
	if(isPushed){
		while(!stack1->isEmpty()){
			stack2->push(stack1->pop());
		}
	isPushed = false;
	}
}
int StackedQueue::push(int data){
	moveToStack1();
	stack1->push(data);
}
int StackedQueue::pop(){
	moveToStack2();
	return stack2->pop();
}
bool StackedQueue::isEmpty(){
	return stack1->isEmpty() && stack2->isEmpty();
		
}
int StackedQueue::peek(){
	moveToStack2();
	return stack2->peek();
}	

int main(){
	StackedQueue* queue = new StackedQueue;
	cout << queue->pop() << endl;
	cout << queue->pop() << endl;
	cout << "is empty?" << endl;
	cout << queue->isEmpty() << endl;
	queue->push(5);
	queue->push(3);
	queue->push(6);
	cout << "is empty?" << endl;
	cout << queue->isEmpty() << endl;
	cout << queue->pop() << endl;
	cout << queue->pop() << endl;
	queue->push(8);
	cout << queue->peek() << endl;
	cout << queue->pop() << endl;
	cout << queue->pop() << endl;
	cout << queue->pop() << endl;
	cout << queue->pop() << endl;

	// Stack* stack = new Stack();
	// stack->push(1);
	// cout << stack->pop() << endl;

	return 0;
}
